clarity/src/vm/callables.rs

```diff
@@ -300,3 +339,216 @@ impl FunctionIdentifier {
+            trait_identifier: trait_id,
+        }) => {
+            assert_eq!(contract_id, &contract_identifier);
+            assert_eq!(trait_id.as_ref().unwrap(), &trait_identifier);
+        }
+        other => panic!("expected Value::CallableContract, got {:?}", other),
+    }
+
+    // (ok principal) -> (ok <trait>)
+    let response_ty =
+        TypeSignature::new_response(trait_ty.clone(), TypeSignature::UIntType).unwrap();
+    let response_contract = Value::okay(contract.clone()).unwrap();
+    let cast_response = implicit_cast(&response_ty, &response_contract).unwrap();
+    let cast_trait = cast_response.expect_result_ok().expect_callable();
+    assert_eq!(&cast_trait.contract_identifier, &contract_identifier);
+    assert_eq!(&cast_trait.trait_identifier.unwrap(), &trait_identifier);
+
+    // (err principal) -> (err <trait>)
+    let response_ty =
+        TypeSignature::new_response(TypeSignature::UIntType, trait_ty.clone()).unwrap();
+    let response_contract = Value::error(contract.clone()).unwrap();
+    let cast_response = implicit_cast(&response_ty, &response_contract).unwrap();
+    let cast_trait = cast_response.expect_result_err().expect_callable();
+    assert_eq!(&cast_trait.contract_identifier, &contract_identifier);
+    assert_eq!(&cast_trait.trait_identifier.unwrap(), &trait_identifier);
+
+    // (list principal) -> (list <trait>)
+    let list_ty = TypeSignature::list_of(trait_ty.clone(), 4).unwrap();
+    let list_contract = Value::list_from(vec![contract.clone(), contract2.clone()]).unwrap();
+    let cast_list = implicit_cast(&list_ty, &list_contract).unwrap();
+    let items = cast_list.expect_list();
+    for item in items {
+        let cast_trait = item.expect_callable();
+        assert_eq!(&cast_trait.trait_identifier.unwrap(), &trait_identifier);
+    }
+
+    // {a: principal} -> {a: <trait>}
+    let a_name = ClarityName::from("a");
+    let tuple_ty = TypeSignature::TupleType(
+        TupleTypeSignature::try_from(vec![(a_name.clone(), trait_ty.clone())]).unwrap(),
+    );
+    let contract_tuple_ty = TypeSignature::TupleType(
+        TupleTypeSignature::try_from(vec![(a_name.clone(), TypeSignature::PrincipalType)]).unwrap(),
+    );
+    let mut data_map = BTreeMap::new();
+    data_map.insert(a_name.clone(), contract.clone());
+    let tuple_contract = Value::Tuple(TupleData {
+        type_signature: TupleTypeSignature::try_from(vec![(
+            a_name.clone(),
+            TypeSignature::PrincipalType,
+        )])
+        .unwrap(),
+        data_map,
+    });
+    let cast_tuple = implicit_cast(&tuple_ty, &tuple_contract).unwrap();
+    let cast_trait = cast_tuple
+        .expect_tuple()
+        .get(&a_name)
+        .unwrap()
+        .clone()
+        .expect_callable();
+    assert_eq!(&cast_trait.contract_identifier, &contract_identifier);
+    assert_eq!(&cast_trait.trait_identifier.unwrap(), &trait_identifier);
+
+    // (list (optional principal)) -> (list (optional <trait>))
```

Comment:
Can there be a test for `(list principal) -> (list <trait>)`, or is that illegal?  Also, if you're going to check `list`s of `optional`s, then could you also check:

* `list` of `(response <trait> uint)`
* `list` of `(response uint <trait>)`
* `(some (list (optional <trait>)))`
* `(ok (list (optional <trait>)))`
* `(err (list (optional <trait>)))`
...etc.

I think we need to test that you can nest lists inside of optionals and responses while containing traits, in order to convince ourselves that we can put traits inside arbitrarily-nested structures and everything should work.

Reply:
`(list principal) -> (list <trait>)` is above, beginning at line 497.

My thought was that if I tested one of these variants, that should cover the unique parts of the code. I can add some more, but I think the gain is diminishing.

Reply:
More tests added in 64fe8486c